crates/CrateAssembler.kt

```diff
@@ -0,0 +1,213 @@
+import com.eclipsesource.json.JsonArray
+import com.eclipsesource.json.JsonObject
+import com.fasterxml.jackson.dataformat.toml.TomlMapper
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry
+import org.apache.commons.compress.archivers.tar.TarArchiveOutputStream
+import org.apache.commons.compress.utils.IOUtils
+import picocli.CommandLine
+import picocli.CommandLine.Command
+import picocli.CommandLine.Option
+import java.io.BufferedOutputStream
+import java.io.ByteArrayInputStream
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.nio.file.Path
+import java.util.concurrent.Callable
+import java.util.zip.GZIPOutputStream
+import kotlin.system.exitProcess
+
+
+data class CargoTomlPackage(
+    val name: String,
+    val edition: String,
+    val version: String,
+    val authors: List<String>,
+    val homepage: String,
+    val repository: String,
+    val documentation: String,
+    val description: String,
+    val readme: String
+)
+
+data class CargoToml(
+    val `package`: CargoTomlPackage,
+    val dependencies: Map<String, String>
+)
+
+
+@Command(name = "crate-assembler", mixinStandardHelpOptions = true)
+class CrateAssembler : Callable<Unit> {
+    @Option(names = ["--srcs"], split = ";", required = true)
+    lateinit var srcs: Array<File>
+
+    @Option(names = ["--deps"], split = ";", required = true)
+    lateinit var deps: Array<String>
+
+    @Option(names = ["--output-crate"], required = true)
+    lateinit var outputCrateFile: File
+
+    @Option(names = ["--output-metadata-json"], required = true)
+    lateinit var outputMetadataFile: File
+
+    @Option(names = ["--root"], required = true)
+    lateinit var crateRoot: Path
+
+    @Option(names = ["--edition"], required = true)
+    lateinit var edition: String
+
+    @Option(names = ["--name"], required = true)
+    lateinit var name: String
+
+    @Option(names = ["--authors"], split = ";")
+    lateinit var authors: Array<String>
+
+    @Option(names = ["--description"], required = false)
+    lateinit var description: String
+
+    @Option(names = ["--documentation"], required = false)
+    lateinit var documentation: String
+
+    @Option(names = ["--homepage"], required = false)
+    lateinit var homepage: String
+
+    @Option(names = ["--keywords"], split = ";")
+    lateinit var keywords: Array<String>
+
+    @Option(names = ["--categories"], split = ";")
+    lateinit var categories: Array<String>
+
+    @Option(names = ["--license"], required = false)
+    lateinit var license: String
+
+    @Option(names = ["--repository"], required = false)
+    lateinit var repository: String
+
+    @Option(names = ["--readme-file"])
+    var readmeFile: File? = null
+
+    @Option(names = ["--version-file"])
+    lateinit var versionFile: File
+
+    override fun call() {
+        val prefix = "$name-${versionFile.readText()}"
+
+        val libraryRoot = crateRoot.parent.toAbsolutePath()
+        outputCrateFile.outputStream().use { fos ->
```

Comment:
what does the code from line 117 to 141 do? it looks like it correspond to one high-level action. in that case, does it make sense to extract it into its own function? that way we keep `call()` function high level.